File: chess-app/src/engine/pieces/Pawn.js
import Piece from './Piece.js';

export default class Pawn extends Piece {
    constructor(color) {
        super('pawn', color);
    }

    getPseudoLegalMoves(board, fromPos, attack=false){
        let moves = [];
        const movementDir = this.color === 'white' ? -1 : 1;
        const offsets = this.hasMoved ? [[1, 0]] : [[1, 0], [2, 0]];

        if (!attack) {
            offsets.forEach((offset)=>{
                let newPos = fromPos.map((value, index) => value + (offset[index] * movementDir));
                if (board.isInsideBoard(newPos)){
                    let pieceAtNewPos = board.getPieceAt(newPos);
                    if (pieceAtNewPos === null){
                        moves.push(newPos);
                    }
                }
            })
        }

        const captureOffsets = [[1, -1], [1, 1]];
        captureOffsets.forEach((offset)=>{
            let newPos = fromPos.map((value, index) => value + (offset[index] * movementDir));
            if (board.isInsideBoard(newPos)){
                let pieceAtNewPos = board.getPieceAt(newPos);
                if ((pieceAtNewPos !== null && pieceAtNewPos.color !== this.color) 
                    || JSON.stringify(board.enPassantTarget) === JSON.stringify(newPos)) {
                    moves.push(newPos);
                }
            }
        })
        return moves;
    }

    deepCopy() {
        const copy = new Pawn(this.color);
        copy.hasMoved = this.hasMoved;
        return copy;
    }
}

File: chess/src/main/java/com/app/chess/Pieces/Pawn.java
package com.app.chess.Pieces;

import com.app.chess.*;
import java.util.List;
import java.util.Arrays;
import java.util.ArrayList;

public class Pawn extends Piece {
    public Pawn(Player color) {
        super(PieceType.PAWN, color);
    }

    public Pawn() {
        super(PieceType.PAWN);
    }

    @Override
    public List<int []> getPsuedoLegalMoves(Board board, int[] fromPos, boolean attack) {
        List<int[]> legalMoves = new ArrayList<>();
        int direction = (color == Player.WHITE) ? -1 : 1; // White moves up (-1), Black moves down (+1)

        List<int[]> forwardMoves = hasMoved ? List.of(new int[]{direction, 0}) 
                                            : List.of(new int[]{direction, 0}, new int[]{2 * direction, 0});

        if (!attack) {
            for (int[] move : forwardMoves) {
                int[] nextPos = {fromPos[0] + move[0], fromPos[1] + move[1]};
                if (board.isInsideBoard(nextPos) && board.getPieceAt(nextPos) == null) {
                    legalMoves.add(nextPos);
                } else {
                    break; // Stop if blocked
                }
            }
        }

        // Capture moves
        int[][] captureOffsets = {{direction, -1}, {direction, 1}};
        for (int[] offset : captureOffsets) {
            int[] capturePos = {fromPos[0] + offset[0], fromPos[1] + offset[1]};
            if (board.isInsideBoard(capturePos)) {
                Piece targetPiece = board.getPieceAt(capturePos);
                if ((targetPiece != null && targetPiece.color != color)
                    || (Arrays.equals(board.enPassantTarget, capturePos))) {
                    legalMoves.add(capturePos);
                }
            }
        }
        return legalMoves;
    }
    
    @Override
    public Piece deepCopy() {
        Pawn copy = new Pawn(this.color);
        copy.hasMoved = this.hasMoved;
        return copy;
    }
}
